### PR TITLE
Initiatives CRUD Methods

### PR DESCRIPTION
**Pull request by dbouwman**
This adds all the fine-grained fn's that address:

- getting an initiative by id and applying all schema migrations
- activating an initiative
- deleting an initiative
- updating an initiative

Upon release, we can merge a PR for Hub that will remove a bunch of dupe code w/ this.

### Note
I needed the ability to use `project` on the geometry service. Since we don't have an AGRjs module for that, I just added it into the Initiatives module... but we should hoist that over to AGRjs.

### CODE REVIEW

File: packages/initiatives/src/util.ts [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+  // handle jpg
+  if (extension === "jpg") {
+    extension = "jpeg";
+  }
+  return `image/${extension}`;
+}
+
+/**
+ * Load an image from a url, and store it as a resource on an existing item
+ *
```
Comment (jgravois) [inline packages/initiatives/src/util.ts]: easier/better to use [`@esri/arcgis-rest-items/addItemResource()`](https://esri.github.io/arcgis-rest-js/api/items/addItemResource/)?

File: packages/initiatives/test/groups.test.ts [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+      // const createGroupSpy = spyOn(GroupApi, "createGroup").and.callFake(
+      //   (opts: IGroupAddRequestOptions) => {
+      //     return Promise.resolve({ success: true, group: { id: "3ef" } });
+      //   }
+      // );
+      // const protectGroupSpy = spyOn(GroupApi, "protectGroup").and.callFake(
+      //   (opts: IGroupIdRequestOptions) => {
+      //     return Promise.resolve({ success: true });
+      //   }
+      // );
```
Comment (jgravois) [inline packages/initiatives/test/groups.test.ts]: safe to remove

File: packages/initiatives/test/groups.test.ts [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+      // const createGroupSpy = spyOn(GroupApi, "createGroup").and.callFake(
+      //   (opts: IGroupAddRequestOptions) => {
+      //     return Promise.resolve({ success: true, group: { id: "3ef" } });
+      //   }
+      // );
+      // const protectGroupSpy = spyOn(GroupApi, "protectGroup").and.callFake(
+      //   (opts: IGroupIdRequestOptions) => {
+      //     return Promise.resolve({ success: true, id: "3ef" });
+      //   }
+      // );
```
Comment (jgravois) [inline packages/initiatives/test/groups.test.ts]: safe to remove

File: packages/initiatives/test/remove.test.ts [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        expect(fetchInitiativeSpy.calls.count()).toEqual(1);
+        expect(unprotectItemSpy.calls.count()).toEqual(1);
+        expect(removeGroupSpy.calls.count()).toEqual(1);
+        expect(detachSpy.calls.count()).toEqual(0);
+        expect(removeItemSpy.calls.count()).toEqual(1);
+        expect(progressCallbackSpy.calls.count()).toEqual(5);
+        done();
+      })
+      .catch(ex => {
+        done();
```
Comment (jgravois) [inline packages/initiatives/test/remove.test.ts]: fail()

File: packages/sites/src/sites.ts [hunk 5]
```diff
@@ -2,16 +2,15 @@
...
+import { getHubUrl } from "@esri/hub-domains";
 
 /**
  * Fetch the domains associated with a Hub Site.
  * @param siteId - Identifier of the Hub Site
  * @param requestOptions - request options that may include authentication
  * @returns A Promise that will resolve with the domains associated with the site.
  */
-export function fetchDomains(
+export function fetchSiteDomains(
```
Comment (jgravois) [inline packages/sites/src/sites.ts]: this is a breaking name change.

File: packages/sites/src/sites.ts [hunk 6]
```diff
@@ -24,23 +23,23 @@ export function fetchDomains(
...
+ * Fetch the domain associated with a Hub Site. Since a site may have a
+ * custom domain, in addition to a default domain, we will return the
+ * custom domain over the default domain.
+ *
  * @param siteId - Identifier of the Hub Site
  * @param requestOptions - request options that may include authentication
  * @returns A Promise that will resolve with the domains associated with the site.
  */
-export function fetchDomain(
+export function fetchSiteDomain(
```
Comment (jgravois) [inline packages/sites/src/sites.ts]: another breaking name change. (i'm okay with it. just sayin)

File: packages/sites/src/sites.ts [hunk 7]
```diff
@@ -51,34 +50,15 @@ export function fetchDomain(
         return acc;
       }, null);
       if (customEntry) {
+        // return the custom domain
         return customEntry.domain;
       } else {
-        // WAT!?! just pick the first one
+        // just pick the first one
```
Comment (jgravois) [inline packages/sites/src/sites.ts]: i _disapprove_ of this change.

File: packages/solutions/README.md [hunk 8]
```diff
@@ -17,10 +17,71 @@
 npm install @esri/hub-solutions
 ```
 
+## Coarse Grained API
```
Comment (jgravois) [inline packages/solutions/README.md]: README.md > guide > john 😄

File: packages/initiatives/src/geometry.ts [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+      ...requestOptions
+    } as IRequestOptions;
+
+    const gparam = {
+      geometryType,
+      geometries
+    };
+
+    rOptions.params = {
+      geometries: JSON.stringify(gparam),
```
[the remark this answers is not in the document]
Comment (dbouwman) [reply packages/initiatives/src/geometry.ts]: so it does :)

File: packages/initiatives/src/geometry.ts [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    inSR: any,
+    outSR: any,
+    geometryType: string,
+    geometries: [any],
+    requestOptions: IRequestOptions
+  ): Promise<any> {
+    const url = `${serviceUrl}/project`;
+
+    const rOptions = {
+      httpMethod: "POST",
```
[the remark this answers is not in the document]
Comment (dbouwman) [reply packages/initiatives/src/geometry.ts]: 👍

File: packages/initiatives/README.md [hunk 11]
```diff
@@ -17,16 +17,117 @@
 npm install @esri/hub-initiatives
 ```
 
+
+## Coarse Grained API
```
[the remark this answers is not in the document]
Comment (dbouwman) [reply packages/initiatives/README.md]: Moved it into a rough guide...

File: packages/domains/README.md [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+### Example
+
+```bash
+npm install @esri/hub-domains
+```
+
+```js
+import { comingSoon  } from '@esri/hub-domains';
+
+comingSoon()
```
[the remark this answers is not in the document]
Comment (dbouwman) [reply packages/domains/README.md]: done

File: packages/initiatives/src/util.ts (hunk 1 above)
Comment (dbouwman) [reply packages/initiatives/src/util.ts]: I tried to refactor this to use the ARGjs methods, but failed utterly.

File: packages/sites/src/sites.ts (hunk 7 above)
Comment (dbouwman) [reply packages/sites/src/sites.ts]: WAT?!? :)

File: packages/solutions/README.md (hunk 8 above)
Comment (dbouwman) [reply packages/solutions/README.md]: Once we actually *have* solutions code, we can move this... right now it's more a "plan" for the API we want...

File: packages/sites/src/sites.ts (hunk 6 above)
Comment (dbouwman) [reply packages/sites/src/sites.ts]: Not sure why I changed this... reverted...

File: packages/sites/src/sites.ts (hunk 5 above)
Comment (dbouwman) [reply packages/sites/src/sites.ts]: Not sure why I changed this... reverted...

**Comment by jgravois** [comment]
i just pushed a commit that:
1. refactors `addImageAsResource()` to use `addItemResource()` internally
2. removes the last mention of `IInitiative` (in favor of `IInitiativeItem` and `IInitiativeModel`
3. removes a few stray arguments and imports that weren't actually being used.

i tested my changes manually using a simple browser app that references the UMD.

```js
const options = {
    authentication: new arcgisRest.UserSession({
      username: "[insert]",
      password: "[yours]"
    })
  }

  arcgisHub.addImageAsResource(
    "fe34b1",
    "jgravois",
    "headshot-pacifier.jpg",
    "http://edn.maps.arcgis.com/sharing/rest/community/users/jgravois/info/headshot-pacifier.jpg",
    options
  )
    .then(response => console.log(response));
```

File: packages/initiatives/src/util.ts (hunk 1 above)
Comment (jgravois) [reply packages/initiatives/src/util.ts]: done in 138f6af.

**Pull request merged**